bookkeeper-server/src/main/java/org/apache/bookkeeper/client/LedgerMetadata.java

```diff
@@ -279,6 +294,11 @@ public void addEnsemble(long startEntryId, ArrayList<BookieSocketAddress> ensemb
         return ensembles.get(ensembles.headMap(entryId + 1).lastKey());
     }
 
+    @Override
+    public List<BookieSocketAddress> getEnsembleAt(long entryId) {
+        return getEnsemble(entryId);
```

Comment:
return a deep copy of the ensemble. We don't want people to be able to change behaviour by calling getLedgerMetadata.

Reply:
this should be done in a separate task to make LedgerMetadata implementation immutable. This change is more about interfacing existing class. I would prefer not changing any implementation.

Reply:
This is ortogonal to the LedgerMetadata immutability. This PR is exposing bookkeeper client internals, and if we do so, we must do it in a way that people cannot abuse them. It's a simple change, just add: new ArrayList(getEnsemble(entryId); BookieSocketAddress is already immutable. The NavigableSet will require a little more work, but not a huge amount.

Reply:
it is not exposing *new* client internals. the exposure is already there in LedgerHandle. This interface doesn't change anything we have been exposed.

The reason I don't want constructing a new ArrayList is because this list should be constructed using ImmutableList during we construct the LedgerMetadata. it should not be done on individual gets. so this makes sense to be part of making ledger metadata immutable, not be part of an API abstraction. If you feel strong about this, I can spend time on haveing a follow up issue to make ledger metadata immutable.

Reply:
https://github.com/apache/bookkeeper/commit/95373234a273920fadc06742479ddcc4e603fa1b

It's only been exposed this release cycle AFAICT. 

Please don't start on the immutable ledger metadata stuff yet. I already have a partial patch for that from 3+ years ago that does most of the work, but have been holding it off. Big changes like that should be avoided until all yahoo changes are in.

Reply:
yes. in distributedlog, we wrote some classes under `org.apache.bookkeeper.client` to access this package private field. One of the reasons that I am making this public as part of ReadHandle is to be able to access some metadata. so I don't need to keep any hacky code in distributedlog.

Also I marked this class as `LimitedPrivate` and `Unstable` at this moment. It is subject to a change in future. because I want to make it immutable, but not at this moment. when we make it immutable, we convert this as `Public`. Hope this make sense to you.

Reply:
Making this safe now is a trivial change. I don't see a good reason for not doing so. I can do it myself if you want.

Reply:
again my concern is allocation. It is nothing about trivial. also again this is `LimitedPrivate`, I don't see any reason why you want to push a change that is going to change in future.

Reply:
Is this going to be called so much that allocation is a problem? We should only worry about allocation for stuff that gets called with every add or read. 

Anyhow, I'm not going to block this any longer just for this change.